### ISSUE TITLE
Savestate size changes

### ISSUE THREAD

**Issue by GregorR** [libretro/beetle-psx-libretro#94]
mednafen_psx's savestate size changes based presumably on what's loaded at the time. In typical use cases the only real consequence is that RetroArch spits out `RetroArch [libretro WARN] :: warning, save state size has changed` now and again. With netplay or rewinds, this is a catastrophe.

If we're all willing to bite the bullet on differing savestate sizes, then it's time now to adapt rewind and netplay. Otherwise, mednafen_psx should behave like mupen64plus and report the savestate size as the maximum plausible savestate size.

**Comment by simias** [comment]
Doesn't netplay work by diff'ing the savestates anyway? Even if we report a constant size it will still be a problem if stuff "moves around" in the savestate, right? (I have the same problem in rustation).

**Comment by GregorR** [comment]
Netplay never diffs the savestates, no, it just diffs the input between the local and remote peers. So while it needs savestates to be a fixed size, it doesn't depend on that for anything beyond internal rewind and tranferring states to the remote peer.

**Comment by simias** [comment]
Then why can't the netplay code accept savestates with variable sizes? More generally I find the restriction about savestate sizes in libretro a bit annoying, it forces me to "overcommit" savestate size in rustation just to be sure. It's a bit wasteful IMO.

**Comment by GregorR** [comment]
The netplay code absolutely could be adapted to support savestates of variable size, it's just that we're in general not sure what the contract should be. Rewind and netplay currently assume savestates don't change size, and RetroArch complains if they do. One side or the other has to give, and Twinaphex et al are pushing that you should cave first ;)

**Comment by simias** [comment]
Well I think the root of this is in libretro API itself:

```C
/* Returns the amount of data the implementation requires to serialize 
 * internal state (save states).
 * Between calls to retro_load_game() and retro_unload_game(), the 
 * returned size is never allowed to be larger than a previous returned 
 * value, to ensure that the frontend can allocate a save state buffer once.
 */
RETRO_API size_t retro_serialize_size(void);
```

If beetle changes the savestate size mid-game then it's in violation of the libretro API. I still think that the API is overly restrictive here though but again that's an other issue.

**Comment by GregorR** [comment]
I agree, but my concern at present is just with making RetroArch Netplay the super-awesome killer feature it deserves to be :)

**Comment by simias** [comment]
Yeah sure, I'm not really familiar with beetle's implementation of savestates though. I don't know who implemented that. Maybe @rz5 would know?

I guess I could have a look otherwise.

**Comment by rz5** [comment]
I only ported the rsx renderer from Rust to C++. @twinaphex, @Alcaro and @Themaister are the chaps to ping about API complaints.

Looking at the commit history, @Themaister did the initial commit for the mednafen port and @twinaphex has worked on it extensively, they'll know about the savestate implementation.
